# Hacker News

In this project, we are going through data provided by the Hacker News website, started by Y Incubator. Hacker News is popular in the world of technology and start ups that hundreds to thousands of people visit the website to check out the posts. 

What we are interested in are posts that are either **"Ask Hn"** or **"Show HN"** and determining what insights we can get from these posts. The data contains approximately 300,000 rows from the original set and was reduced to about 20,000 for the sake of this project. Below are the descriptions of the data:

- `id`: the unique identifier from Hacker News for the post
- `title`: the title of the post
- `url`: the URL that the posts links to, if the post has a URL
- `num_points`: the number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
- `num_comments`: the number of comments on the post
- `author`: the username of the person who submitted the post
- `created_at`: the date and time of the post's submission

## Importing the files and libraries

In [4]:
from csv import reader

opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)

# Removed & separated the headers from the list of lists
headers = hn[0]
hn = hn[1:] 

# View the first 5 rows of data from the CSV file
print(hn[:5])

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


In [2]:
headers

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

## Data Cleaning & Extraction

We want to focus on title posts that start with Ask HN or Show HN. In order to find these posts, we'll use the string method `startswith` to determine a boolean as we iterate through each title. 

### Extracting Ask HN, Show HN, and Other posts into separate lists

In [26]:
# Initialize 3 empty lists
ask_posts, show_posts, other_posts = [], [], []

# Iterate through each row and determine the title column to a variable
for row in hn:
    title = row[1]
    
    # Conditional statement: if ask hn, if show hn, else other
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

# We will check the length of the lists
print("Ask post total:", len(ask_posts))
print('Show post total:', len(show_posts))
print('Other post total:', len(other_posts))

Ask post total: 1744
Show post total: 1162
Other post total: 17194


In [27]:
# Viewing the first 5 posts of each list
print(ask_posts[:5])
print(show_posts[:5])
print(other_posts[:5])

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]
[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 'h

## Analyzing the data
---

### Calculating the average amount of comments between Ask HN and Show HN posts

In [31]:
# Find Ask HN post comment average
total_ask_comments = 0

for post in ask_posts:
    total_ask_comments += int(post[4])
    
avg_ask_comments = total_ask_comments/len(ask_posts)

# And then for Show HN post
total_show_comments = 0

for post in show_posts:
    total_show_comments += int(post[4])
    
avg_show_comments = total_show_comments/len(show_posts)

print('Average comment per ask post:', avg_ask_comments)
print('Average comment per show post:', avg_show_comments)

# Lastly, for giggles, Other posts
total_other_comments = 0
for post in other_posts:
    total_other_comments += int(post[4])
avg_other_comments = total_other_comments/len(other_posts)
print('Average comment per other post:', avg_other_comments)

Average comment per ask post: 14.038417431192661
Average comment per show post: 10.31669535283993
Average comment per other post: 26.8730371059672


We discover that the average comment per show post (*10.32*) is less than the average comment per ask post (*14.0*). Although that average comment per other post is significantly higher (*26.9*), that is not the main focus of this project.

One observation is that in an "Ask HN" post, people who are posting these posts are looking for answers. An answer, comment, or reply, is an expected thing to come out of these posts. There also seems to be a lot of people asking considering the total number of these posts is higher than the "Show HN" posts.

### Calculating the number of posts and comments in each hour per day

We will determine if ask posts created at a certain time are more likely to attract comments for Ask HN posts. We will most likely do the same for Show and Other posts.

In [32]:
import datetime as dt

In [36]:
## Ask HN ##
# Create an empty list to contain a list of lists
result_list = []

# Iterate over ask posts
for post in ask_posts:
    created_at = post[6]
    num_comments = int(post[4])
    # Append a list of these two elements into result list
    result_list.append([created_at, num_comments])

counts_by_hour = {}
comments_by_hour = {}
for item in result_list:
    # Extracting the hour from the date
    date = dt.datetime.strptime(item[0], "%m/%d/%Y %H:%M")
    hour = date.strftime("%H")
    # Determining counts by hour and comments by hour through conditional if statement
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = item[1]
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += item[1]

In [42]:
print('Number of posts posted by the hour\n', counts_by_hour, '\n')
print('Number of comments posted by hour\n',comments_by_hour)

Number of posts posted by the hour
 {'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58} 

Number of comments posted by hour
 {'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


We will use the dictionaries we created to calculate the average number of comments for posts created during each hour of the day.

In [46]:
avg_by_hour = []

for hour in comments_by_hour:
    avg_by_hour.append([hour, comments_by_hour[hour] / counts_by_hour[hour]])
    
avg_by_hour

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

We will sort the list of lists and print the five highest values in a format that's easier to read.

In [48]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    
print(swap_avg_by_hour)

[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


In [53]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
sorted_swap

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [55]:
print('Top 5 Hours for Ask Posts Comments')
for row in sorted_swap[:5]:
    # Used .strptime to return datetime obj, then .strftime to specify the format of the time
    hour = dt.datetime.strptime(row[1], "%H")
    hour_formatted = hour.strftime("%H:%M")
    # We also rounded the decimals
    print("{}: {:.2f} average comments per post".format(hour_formatted,row[0]))

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


According to our analysis, posting around 3PM will get the most comments/replies to an Ask HN post. There also seems to be a high amount of comments being posted on Ask HN posts around 2AM and 10PM, the 2AM must be due to possibly a comment coming from someone where it is there 3PM over there. The same concept must said about the 10PM time slot because maybe the people who are reply might be people in the same time zone as you, just getting off work and having time to reply before going to bed. 

If the the algorithm about showing the most newest/relevant posts on Hacking News to the community is very exceptional, then these times of 3PM, 2AM, and 10PM are the best times to post an Ask HN post.

### Let's calculate the same numbers for Show HN and Others posts.

In [ ]:
## Show HN ##
